Include the position information for the Expression nodes in the Fhipath Parser

brianpos opened the pull request:
## Description
The Expression class has a `Location` property which is intended to store the position information of each Expression object in the AST. This has never been populated (was expected to go via the constructor).
This PR uses the sprache SetPos/Positioned() capability to inject that information while parsing.
The old anticipated approach has been retained, and leveraged for unit testing to manually populate the data.

It also adds support for skipping over comments in expressions.

## Related issues
Resolves #2517

## Testing
There are many new unit tests to verify the positioning information.

## FirelyTeam Checklist
- [ ] **Update the title** of the PR to be succinct and less than 50 characters
- [ ] Mark the PR with the label **breaking change** when this PR introduces breaking changes

brianpos committed:
Introduce a CustomExpression that can permit the addition of other expression types without impacting visitor implementations

brianpos committed:
Add a debugger display for function expressions

brianpos committed:
Add a new Expresion type for Identifier (IdentifierExpression). Derives from Constant to retain compatibility with existing code that might be using that type.

brianpos committed:
Introduce a BracketExpression type as a CustomExpression to permit round tripping more easily.
This will reduce to nothing if not handled in a visitor

brianpos committed:
Set the location property while parsing the expression (using the Sprache Position information)

brianpos committed:
* Roundtrip Visualizer to output a canonical version of the expression (standardizing whitespace)
* Brackets can handle having extra whitespace around them
* Brackets support left/right subtokens inside
* FunctionCalls support left/right subtokens inside
* Position information updated for function calls (and child expressions too)
* Roundtripping fhirpath expressions unit tests

brianpos committed:
Merge branch 'develop' into feature/fhirpath-parsing

brianpos committed:
#2517 line numbering information while parsing the expressions into the Expression objects
and also handling comments in fhirpath expressions (they are skipped and not retained in the expression AST)

brianpos commented on src/Hl7.Fhir.Base/FhirPath/Parser/Grammar.cs:
```diff
@@ -65,47 +65,51 @@ private static IResult<P.Quantity> quantityParser(IInput i)
             Literal
             .Or(FunctionInvocation(AxisExpression.That))
-            .Or(Lexer.ExternalConstant.Select(n => BuildVariableRefExpression(n))) //Was .XOr(Lexer.ExternalConstant.Select(v => Eval.ExternalConstant(v)))
+            .XOr(Lexer.ExternalConstant.Select(n => BuildVariableRefExpression(n)).Positioned()) //Was .XOr(Lexer.ExternalConstant.Select(v => Eval.ExternalConstant(v)))
```
@Kasdejong @ewoutkramer This is the line that I was referring to in Slack.

ewoutkramer replied:
This was changed from Xor to Or because we shortly had an additional grammar rule for environment variables starting with `%%` - so to distinguish between an external constant and an environment variable, we had to do a full lookahead. That bit got removed, so I think it should all be Xors, but I'll wait for @Kasdejong to confirm.

Kasdejong replied:
I am testing this right now, no errors yet when removing it. I'm pretty sure this can be XOR again!

brianpos replied:
It is now Xor and all tests are working, so I'll assume this is ok now.

brianpos commented:
I need to recheck this against my other branch... fhirpath-position (which has the round-trip visualizer).
https://github.com/FirelyTeam/firely-net-sdk/compare/feature/fhirpath-parsing...feature/fhirpath-position
Will do that tomorrow.

brianpos committed:
Merge branch 'develop' into feature/fhirpath-position

brianpos committed:
Minor compilation issues

brianpos committed:
Roundtrip visitor and tests from the position branch

brianpos committed:
Merge branch 'feature/fhirpath-position' into feature/fhirpath-parsing

brianpos committed:
Change name from RoundtripVisitor to CanonicalVisitor

brianpos commented:
Branch is now good for review.
The other changes I have locally are for capturing the whitespace/comments into subtokens.
And I'd also like to change the name from `SubTokenExpression` to just `SubToken`

brianpos committed:
Rename SubTokenExpression to just SubToken

ewoutkramer committed:
Merge branch 'develop' into feature/fhirpath-parsing

ewoutkramer commented on src/Hl7.Fhir.Base/FhirPath/Expressions/CanonicalVisitor.cs:
```diff
@@ -0,0 +1,207 @@
+
+namespace Hl7.FhirPath.Expressions
+{
+    public class CanonicalVisitor : ExpressionVisitor<StringBuilder>
```
We're trying to (sometimes retrospectively) document all public classes - can you add documentation here?

brianpos replied:
Documented now

ewoutkramer commented on src/Hl7.Fhir.Base/FhirPath/Expressions/ExpressionNode.cs:
```diff
@@ -27,26 +30,149 @@ protected Expression(TypeSpecifier type)
             ExpressionType = type;
         }
 
+        // [Obsolete("This will be removed... this happens through the SetPos()", false)]
```
Do we need to make this Obsolete indeed?

brianpos replied:
I've retained the constructor for both backward compatibility, and also the unit tests use them quite conveniently too.
So I guess just make a decision that they can remain and remove the comment...

ewoutkramer commented on src/Hl7.Fhir.Base/FhirPath/Expressions/ExpressionNode.cs:
```diff
@@ -10,14 +10,17 @@
 namespace Hl7.FhirPath.Expressions
 {
-    public abstract class Expression : IEquatable<Expression>
+    public abstract class Expression : IEquatable<Expression>, Sprache.IPositionAware<Expression>
```
I know we discussed whether this was breaking compile compatibility, but seem to remember we thought is was ok. Do you still remember the arguments?

brianpos replied:
Correct, the breaking part that I had in an earlier diff was that I removed the `String location = null` constructors. These have all been put back (correct me if you find them)
If someone has added their own expressions, then they just wouldn't get position information during parsing (which they'd have to have messed with themselves too)

So I'm pretty sure we;re good here.

brianpos committed:
Equals overloads are not accurate and not actually being used in many of the unit tests.

ewoutkramer commented on src/Hl7.Fhir.Base/FhirPath/Expressions/ExpressionNode.cs:
```diff
@@ -27,26 +30,149 @@ protected Expression(TypeSpecifier type)
+            // Also check the location information if it is present in the other
+            // This is the new functionality added for the unit testing
+            // (previously location data was never present, and thus never compared)
+            if (Location != null)
```
If this node has no location but the other has, are they considered equal?

brianpos replied:
Yes, didn't want to change the pre-existing functionality.
The way this is will permit a test that DOES provide position information to check it, but if not is still fine.

ewoutkramer commented on src/Hl7.Fhir.Base/FhirPath/Expressions/ExpressionNode.cs:
```diff
@@ -27,26 +30,149 @@ protected Expression(TypeSpecifier type)
+
 
-    public class ConstantExpression : Expression
+    public class SubTokenExpression : IPositionAware<SubTokenExpression>
```
Isn't this a CustomExpression? But what would it reduce to?

ewoutkramer replied:
ACtually, does this need to be an expression at all?  Or can the parsers for the braces return a `record` with the necessary data?

brianpos replied:
I'm happy this new SubToken class is not an expression, as it doesn't fit in the normal expression tree for evaluation.
Unlike the custom expressions, which do form a part of the tree.

brianpos replied:
I've moved all the SubToken stuff into it's own file now, and included documentation too, so I think this is now all clear.
And the description of what is an Expression, SubToken and CustomExpression.

ewoutkramer commented on src/Hl7.Fhir.Base/FhirPath/Expressions/ExpressionNode.cs:
```diff
@@ -27,26 +30,149 @@ protected Expression(TypeSpecifier type)
+
 
-    public class ConstantExpression : Expression
+    public class SubTokenExpression : IPositionAware<SubTokenExpression>
```
Can you document this public class?

brianpos replied:
Documentation now included

brianpos committed:
Correct the expected unit test values.

ewoutkramer commented on src/Hl7.Fhir.Base/FhirPath/Expressions/ExpressionNode.cs:
```diff
@@ -27,26 +30,149 @@ protected Expression(TypeSpecifier type)
+        IdentifierExpression IPositionAware<IdentifierExpression>.SetPos(Position startPos, int length) => SetPos<IdentifierExpression>(startPos, length);
+    }
+
+    // Discuss: Reduce function? - Skip this in the compile stage? - Update the visitor to skip the bracket too
```
Reducing this to Operand seems to be doing precisely that!  (or maybe you mean the compiler does not call reduce() yet - that could be true).

brianpos replied:
This was a comment from doing our review way back.
I believe that it should be good now.
I verified that compile does indeed reduce down properly too.

brianpos replied:
So I'm ok if we just delete the comment if you're also good with it as is.

ewoutkramer commented on src/Hl7.Fhir.Base/FhirPath/Expressions/ExpressionVisitor.cs:
```diff
@@ -19,5 +19,10 @@ public abstract class ExpressionVisitor<T>
+
+        public virtual T VisitCustomExpression(CustomExpression expression)
+        {
+            return expression.Reduce().Accept<T>(this);
```
Ah, here it is ;-)

brianpos replied:
And also added some extra documentation

ewoutkramer commented on src/Hl7.Fhir.Base/FhirPath/Expressions/FhirPathExpressionLocationInfo.cs:
```diff
@@ -0,0 +1,40 @@
+namespace Hl7.FhirPath.Expressions
+{
+    [DebuggerDisplay(@"\{{DebuggerDisplay,nq}}")]
+    public class FhirPathExpressionLocationInfo : ISourcePositionInfo
```
Please add documentation for public members + nullability annotations.

brianpos replied:
Documentation added

ewoutkramer reviewed (changes requested):
There's a risk I am making comments about things we discussed before at the WGM, but it was so long ago, that I have forgotten our conclusions ;-)

In general, we should now be adding doccomments to any (new) public classes and other surface elements, so that's my major feedback.  If possible, adding `#nullable enable` and specifying nullability markers would be nice, but sometimes that is a pandora's box.

brianpos committed:
Lots of code comments/documentation for the public classes/properties
Inclusion of an EchoVisitor (for future use)
Move the Sub-Token class into it's own file

brianpos committed:
Missed some future parsing use properties (used in the EchoVisitor)
And more code comments.

brianpos committed:
Remove the commented out Lambda Expression as it is not a part of the fhirpath language

brianpos committed:
Additional code comments

brianpos commented:
I've completed all the changes requested.
If you want to also see the comment/whitespace parsing changes, they are in this branch
https://github.com/FirelyTeam/firely-net-sdk/tree/feature/fhirpath-parsing-comments

Diff here:
https://github.com/FirelyTeam/firely-net-sdk/compare/feature/fhirpath-parsing...feature/fhirpath-parsing-comments
_Note that the TokenIgnoreComments that I introduced here has been removed - which is added in the first pass in this PR - so isn't a change to the public interface when combined._

GinoCanessa commented on src/Hl7.Fhir.Base/FhirPath/Expressions/SubToken.cs:
```diff
@@ -0,0 +1,139 @@
+        /// <summary>
+        /// Any leading whitespace or comments encountered immediately before this SubToken
+        /// </summary>
+        public IEnumerable<WhitespaceSubToken> leadingWhitespace { get; internal set; }
```
Should this be Pascal Cased?

ewoutkramer replied:
Yes.

GinoCanessa commented on src/Hl7.Fhir.Base/FhirPath/Expressions/SubToken.cs:
```diff
@@ -0,0 +1,139 @@
+        /// <summary>
+        /// Any trailing whitespace or comments encountered immediately after this SubToken
+        /// </summary>
+        public IEnumerable<WhitespaceSubToken> trailingWhitespace { get; internal set; }
```
Should this be Pascal Cased?

brianpos replied:
Resolved in the other branch when that is merged in too.

brianpos commented on src/Hl7.Fhir.Base/FhirPath/Expressions/EchoVisitor.cs:
```diff
@@ -0,0 +1,301 @@
+                    }
+                    break;
+                case "String":
+                    _result.Append($"'{t}'");
```
@ewoutkramer ?
This line should really be handling the delimiting of the strings too.
Should the `Functions.StringOperators.EscapeJson` be the appropriate function to use there?
```
_result.Append("'" + Functions.StringOperators.EscapeJson(t) + "'"); 
```
Note: yes something does need to be done here, I found this round tripping all the search parameters/invariants.

ewoutkramer replied:
Yeah, if the goal of this visitor is to roundtrip to the original fhirpath as much as possible, it should escape stuff. Not easy to really roundtrip (as you cannot distinguish between allowed unicode characters and escaped unicodes anymore), but that would be good enough.  So, it should "undo" the parsing of the string.

If this were C# we could leverage Roslyn:

```
  return SyntaxFactory.LiteralExpression(SyntaxKind.StringLiteralExpression, SyntaxFactory.Literal(input)).ToFullString();
```

But I don't know if that's good enough.

ewoutkramer committed:
Merge branch 'develop' into feature/fhirpath-parsing

ewoutkramer commented on src/Hl7.Fhir.Base/FhirPath/Expressions/EchoVisitor.cs:
```diff
@@ -0,0 +1,301 @@
+                case "Integer":
+                    _result.Append($"{t}");
+                    break;
+                case "Long":
```
You could combine all cases that do the same thing, to make clear they are doing the same thing!

ewoutkramer commented:
@brianpos - note that both Gino and I had some review remarks.

Build is currently failing because of a QA step: https://dev.azure.com/firely/firely-net-sdk/_build/results?buildId=38380&view=logs&j=558a83a0-5585-5f85-5265-cdf5febe3f62&t=a961ae7d-38e2-5a82-9685-674f5235ed2d, we have added new constructors which breaks binary compat.  We're not striving to get to binary compat, and I think it's alright for compile compatibility. But I don't remember how to disable the message. @Kasdejong ?

brianpos commented:
Gino's concerns are in the second branch that follows this one.
If you're good with those other doffs I can just bring them all into this PR.
And I thought I'd addressed all your changes. Left the comment so you coukd verify.

I'll have another look at the constructors listed in the build.

brianpos commented:
Should I just merge in the whitespace and comments capturing changes into this branch too now?

brianpos commented:
> @brianpos - note that both Gino and I had some review remarks.
> 
> Build is currently failing because of a QA step: https://dev.azure.com/firely/firely-net-sdk/_build/results?buildId=38380&view=logs&j=558a83a0-5585-5f85-5265-cdf5febe3f62&t=a961ae7d-38e2-5a82-9685-674f5235ed2d, we have added new constructors which breaks binary compat. We're not striving to get to binary compat, and I think it's alright for compile compatibility. But I don't remember how to disable the message. @Kasdejong ?

I've reviewed those messages, It looks like its just that it wants the constructors that have the default location = null parameter split into seperate ones.
I can do that very easily - and have done locally in the other branch ready to push with your approval.

ewoutkramer commented:
> Should I just merge in the whitespace and comments capturing changes into this branch too now?

Nah, let's do two PRs, this one is done & reviewed now.  Just need to either split those constructors, or wait for Kas or Marten to tune that exception file.  I could probably find out how, but after this FMG meeting I'll join them in going a short Ascension break, so it will be Monday before we get to it.

brianpos committed:
Change the constructors split out the default ones (which should be binary compatible with the previous release)

(cherry picked from commit f13e4097dc9dc8a8bd8c025adc90aa40f9749040)

ewoutkramer committed:
Merge branch 'develop' into feature/fhirpath-parsing
